[Workers] Add docs for Vitest testing (#13092)
cloudflare/cloudflare-docs: content/workers/testing/vitest-integration/configuration.md

        await append("three");
        expect(await get()).toStrictEqual([
          // All `beforeAll()`s run before `beforeEach()`s
          "all", "describe all", "each", "describe each", "three"
        ]);
      });
      // `append("each")`, `append("describe each")` and `append("three")` undone
      test("four", async () => {
        await append("four");
        expect(await get()).toStrictEqual([
          "all", "describe all", "each", "describe each", "four"
        ]);
      });
      // `append("each")`, `append("describe each")` and `append("four")` undone
    });
    ```

    </details>

- {{<code>}}singleWorker: {{<type>}}boolean{{</type>}}{{<prop-meta>}}optional{{</prop-meta>}}{{</code>}} — Defaults to `false`

  - Runs all tests in this project serially in the same Worker, using the same module cache. This can significantly speed up execution if you have lots of small test files. Refer to the [Isolation and concurrency](/workers/testing/vitest/internal-details/) page for more information on the isolation model.

- {{<code>}}miniflare: {{<type>}}SourcelessWorkerOptions & { workers?: WorkerOptions[]; }{{</type>}}{{<prop-meta>}}optional{{</prop-meta>}}{{</code>}}

>   - Miniflare sourceless-[`WorkerOptions`](https://github.com/cloudflare/workers-sdk/tree/main/packages/miniflare#interface-workeroptions) for configuring bindings and compatibility settings. Use the `main` option above to configure the entrypoint, instead of the Miniflare `script`, `scriptPath`, or `modules` options.<br><br>

Reviewer: ```suggestion
  - Miniflare sourceless-[`WorkerOptions`](https://github.com/cloudflare/workers-sdk/tree/main/packages/miniflare#interface-workeroptions) for configuring bindings and compatibility settings. Use the `main` option above to configure the entry point, instead of the Miniflare `script`, `scriptPath`, or `modules` options.<br><br>
```

>     If your project makes use of multiple Workers, you can configure _auxiliary Workers_ that run in the same `workerd` process as your tests and can be bound to. Auxiliary Workers are configured using the `workers` array, containing regular Miniflare [`WorkerOptions`](https://github.com/cloudflare/workers-sdk/tree/main/packages/miniflare#interface-workeroptions) objects. Note that unlike the `main` Worker, auxiliary Workers:

Reviewer: ```suggestion
    If your project makes use of multiple Workers, you can configure auxiliary Workers that run in the same `workerd` process as your tests and can be bound to. Auxiliary Workers are configured using the `workers` array, containing regular Miniflare [`WorkerOptions`](https://github.com/cloudflare/workers-sdk/tree/main/packages/miniflare#interface-workeroptions) objects. Note that unlike the `main` Worker, auxiliary Workers:
```

>     - Cannot have TypeScript entrypoints—you must compile them to JavaScript first
>     - Use regular Workers module resolution semantics—refer to the [Isolation and concurrency](/workers/testing/vitest/internal-details/#modules) page for more information
>     - Cannot access the `cloudflare:test` module
>     - Do not require specific compatibility dates or flags
>     - Can be written with the [Service Worker syntax](/workers/reference/migrate-to-module-workers/#service-worker-syntax)
>     - Are not affected by global mocks defined in your tests

Reviewer: ```suggestion
    - Cannot have TypeScript entrypoints. You must compile auxiliary Workers to JavaScript first.
    - Use regular Workers module resolution semantics. Refer to the [Isolation and concurrency](/workers/testing/vitest/internal-details/#modules) page for more information.
    - Cannot access the `cloudflare:test` module.
    - Do not require specific compatibility dates or flags.
    - Can be written with the [Service Worker syntax](/workers/reference/migrate-to-module-workers/#service-worker-syntax).
    - Are not affected by global mocks defined in your tests.
```

> - {{<code>}}wrangler: {{<type>}}{ configPath?: string; }{{</type>}}{{<prop-meta>}}optional{{</prop-meta>}}{{</code>}}

Reviewer: ```suggestion
- {{<code>}}wrangler: {{<type>}}{ configPath?: string; }{{</type>}}{{</code>}}{{<prop-meta>}}optional{{</prop-meta>}}
```

>   - Path to Wrangler configuration file to load `main`, compatibility settings and bindings from. These options will be merged with the `miniflare` option above, with `miniflare` values taking precedence. For example, if your Wrangler configuration defined a service binding named `SERVICE` to a worker named `service`, but you included `serviceBindings: { SERVICE(request) { return new Response("body"); } }` in the `miniflare` option, all requests to `SERVICE` in tests would return `body`. Note `configPath` accepts both `.toml` and `.json` files.

Reviewer: ```suggestion
  - Path to Wrangler configuration file to load `main`, compatibility settings and bindings from. These options will be merged with the `miniflare` option above, with `miniflare` values taking precedence. For example, if your Wrangler configuration defined a [service binding](/workers/configuration/bindings/about-service-bindings/) named `SERVICE` to a Worker named `service`, but you included `serviceBindings: { SERVICE(request) { return new Response("body"); } }` in the `miniflare` option, all requests to `SERVICE` in tests would return `body`. Note `configPath` accepts both `.toml` and `.json` files.
```

{{</definitions>}}

{{<Aside type="warning">}}

> You must define a compatibility date of `2022-10-31` or higher, and include `nodejs_compat` in your compatibility flags to use the Workers Vitest integration.

Reviewer: ```suggestion
You must define a compatibility date of `2022-10-31` or higher, and include [`nodejs_compat` in your compatibility flags](/workers/configuration/compatibility-dates/#nodejs-compatibility-flag) to use the Workers Vitest integration.
```

{{</Aside>}}

## `WorkersPoolOptionsContext`

{{<definitions>}}

- {{<code>}}inject: {{<type>}}typeof import("vitest").inject{{</type>}}{{</code>}}

>   - The same `inject()` function usually imported from the `vitest` module inside tests. This allows you to define `miniflare` config based on injected values from [`globalSetup`](https://vitest.dev/config/#globalsetup) scripts. Use this if you have a value in your configuration that is dynamically generated and only known at runtime of your tests. For example, a global setup script might start an upstream server on a random port. This port could be `provide()`d and then `inject()`ed in the configuration for an external service binding or Hyperdrive.

Reviewer: ```suggestion
  - The same `inject()` function usually imported from the `vitest` module inside tests. This allows you to define `miniflare` configuration based on injected values from [`globalSetup`](https://vitest.dev/config/#globalsetup) scripts. Use this if you have a value in your configuration that is dynamically generated and only known at runtime of your tests. For example, a global setup script might start an upstream server on a random port. This port could be `provide()`d and then `inject()`ed in the configuration for an external service binding or [Hyperdrive](/hyperdrive/).
```

An example configuration would be:

```ts
---
filename: vitest.config.ts
---
import { defineWorkersConfig } from "@cloudflare/vitest-pool-workers/config";

export default defineWorkersConfig({
  test: {
    poolOptions: {
      workers: defineWorkersPoolOptions({
        isolatedStorage: true,
        wrangler: {
          configPath: "./wrangler.toml"
        },
      }),
    },
  },
});
```

{{<Aside type="note">}}

Use the [`poolMatchGlobs` option](https://vitest.dev/config/#poolmatchglobs-0-29-4) to only use the Workers Vitest integration for certain tests.

{{</Aside>}}

{{<Aside type="warning">}}

Custom `environment`s or `runner`s are not supported when using the Workers Vitest integration.

{{</Aside>}}

## Functions

The following functions are exported from the `@cloudflare/vitest-pool-workers/config` module.

{{<definitions>}}

> - {{<code>}}defineWorkersPoolOptions(options:{{<param-type>}}WorkersPoolOptions | ((ctx: WorkersPoolOptionsContext) => WorkersPoolOptions | Promise\<WorkersPoolOptions>){{</param-type>}}){{</code>}}

Reviewer: ```suggestion
- {{<code>}}defineWorkersPoolOptions(options:{{<param-type>}}WorkersPoolOptions | ((ctx: WorkersPoolOptionsContext){{</code>}}: WorkersPoolOptions | Promise\<WorkersPoolOptions>){{</param-type>}})
```

---
title: Configuration
pcx_content_type: reference
weight: 4
---

# Configuration

The Workers Vitest integration provides additional configuration on top of Vitest's usual options.
To configure the integration, use the `poolOptions.workers` key.
Use the `defineWorkersPoolOptions()` function from the `@cloudflare/vitest-pool-workers/config` module for type checking and completions.

An example configuration would be:

```ts
---
filename: vitest.config.ts
---
import { defineWorkersConfig } from "@cloudflare/vitest-pool-workers/config";

export default defineWorkersConfig({
  test: {
    poolOptions: {
      workers: defineWorkersPoolOptions({
        isolatedStorage: true,
        wrangler: {
          configPath: "./wrangler.toml"
        },
      }),
    },
  },
});
```

{{<Aside type="note">}}

Use the [`poolMatchGlobs` option](https://vitest.dev/config/#poolmatchglobs-0-29-4) to only use the Workers Vitest integration for certain tests.

{{</Aside>}}

{{<Aside type="warning">}}

Custom `environment`s or `runner`s are not supported when using the Workers Vitest integration.

{{</Aside>}}

## Functions

> The following functions are exported from the `@cloudflare/vitest-pool-workers/config` module.

Reviewer: I've never seen any Cloudflare runtime API documentation where we had a `:` symbol next to the option. `:` should denote a return: https://developers.cloudflare.com/workers/runtime-apis/web-crypto/#subtlecrypto-methods and `<>` will denote wat the Promise is returning.

Is this intentional? What is it representing?
